clippy_lints/src/mut_key.rs

```diff
@@ -103,26 +103,39 @@ fn check_sig<'tcx>(cx: &LateContext<'tcx>, item_hir_id: hir::HirId, decl: &hir::
 fn check_ty<'tcx>(cx: &LateContext<'tcx>, span: Span, ty: Ty<'tcx>) {
     let ty = ty.peel_refs();
     if let Adt(def, substs) = ty.kind() {
-        if [sym::hashmap_type, sym::BTreeMap, sym::hashset_type, sym::BTreeMap]
+        let is_map_type = [sym::hashmap_type, sym::BTreeMap, sym::hashset_type, sym::BTreeMap]
             .iter()
-            .any(|diag_item| cx.tcx.is_diagnostic_item(*diag_item, def.did))
-            && is_mutable_type(cx, substs.type_at(0), span)
-        {
+            .any(|diag_item| cx.tcx.is_diagnostic_item(*diag_item, def.did));
+        if is_map_type && is_mutable_type(cx, substs.type_at(0), span, true) {
             span_lint(cx, MUTABLE_KEY_TYPE, span, "mutable key type");
         }
     }
 }
 
-fn is_mutable_type<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>, span: Span) -> bool {
+fn is_mutable_type<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>, span: Span, is_top_level_type: bool) -> bool {
     match *ty.kind() {
-        RawPtr(TypeAndMut { ty: inner_ty, mutbl }) | Ref(_, inner_ty, mutbl) => {
-            mutbl == hir::Mutability::Mut || is_mutable_type(cx, inner_ty, span)
+        RawPtr(TypeAndMut { ty: inner_ty, mutbl }) => {
+            if is_top_level_type {
+                // Raw pointers are hashed by the address they point to, not what is pointed to.
+                // Therefore, using a raw pointer to any type as the top-level key type is OK.
+                // Using raw pointers _in_ the key type is not, because the wrapper type could
+                // provide a custom `impl` for `Hash` (which could deref the raw pointer).
```

Comment:
I'm not sure the "top-level" disambiguation is necessary.

> the wrapper type could provide a custom `impl` for `Hash` (which could deref the raw pointer).

Do you have an example of this? It seems unlikely/wrong to me to have a specialized `impl Hash` that derefs the pointer. I'm pretty sure this would not happen for any primitive (like array) or std types. Also I think it's okay to lean on the side of having false negatives rather than false positives.

Reply:
> Do you have an example of this?

Surely `Vec` is an example - it contains a raw pointer to some data (contained inside a `Unique<T>`), and its `impl Hash` depends on that data.  So, currently, `HashMap::<Cell<usize>>, ()>` triggers the lint but `HashMap::<Vec<Cell<usize>>, ()>` doesn't (despite being equally dangerous).  This makes sense, because the lint's case for `Adt`s doesn't truly check for mutability - it instead checks for the [`Freeze` trait](https://stdrs.dev/nightly/x86_64-unknown-linux-gnu/core/marker/trait.Freeze.html) which only checks for `UnsafeCell`s **without** indirection.  This also makes sense because (as I understand it) the `Freeze` trait simply says _"the bytes of this type can't change without a mutable reference"_.

So, in fact, I think the ADT clause should recurse over the (fully qualified) types of its fields (i.e. an ADT is mutable iff any of its subtypes are).  Not sure if that's worthy of another PR (or if it's even possible :sweat_smile:).

Reply:
`HashMap<&[*mut Cell<usize>], ()>` or `HashMap<Vec<*mut Cell<usize>>, ()>` should not lint because `impl Hash for Vec<T>` or `&[T]` does not deref the pointer in the generic type.

For the ADT case, what we could do is check if the type is a known std collection (any type where generic parameters are hashed), and if it is, recurse over generic types. This would fix the `Vec<Cell<..>>` case.

Reply:
Ah yes I see this is more complicated than I first thought :sweat_smile:

> For the ADT case, what we could do is check if the type is a known std collection (any type where generic parameters are hashed), and if it is, recurse over generic types. This would fix the `Vec<Cell<..>>` case.

I like this solution - if someone has an ADT with some funky pointer magic and implements `Hash` by (unsafely) dereferencing that pointer, then I think it's justified for Clippy to not lint.  But equally it's worth making the lint as good as possible within performance/dev time/complexity constraints, so I think this is a good middle line.

I'll probs get on to this tomorrow, since I'm out of time for this evening...

Reply:
Right, this has been done!